### PR TITLE
Added some fixes. See description

### LINKED ISSUE #1838: on click of refresh event
Hi, We need to handle the onclick of refresh event, could you please let us know how to handle this , for example, when user clickc on refresh button on table,  
`
$('#tableInstance').bootstrapTable({
}).on('refresh', function (e, data) {
//some logic here
   })`

### LINKED ISSUE #2005: there is an error about export when sidePagination is server
when sidePagination is server and  exportDataType is selected,it will export the datas that i choose basic.
so i check the code in bootstrap-table-export.js and rewrite.

i change 
``` javascript
 that.load(selectedData);
 doExport();
 that.load(data);
```
to
```javascript
  if(that.options.sidePagination === "server"){
     that.refreshOptions({sidePagination:"client"});
     that.load(selectedData);
      doExport();
      that.refreshOptions({sidePagination:"server"});
      that.load(data);
    }else {
      that.load(selectedData);
      doExport();
      that.load(data);
    }
```
in $menu.find('li').click

the result will be right

### LINKED ISSUE #2071: filterControl+height options causes caret bug on Internet Explorer
When using filter control extension, a specific test case can lead to caret issues when filtering rows.

Here is the demo: https://jsfiddle.net/7t5tocn7/2/

Side note: `height` option must be specified for this test case to fail.

Here's the code for reference:

```html
<h3>Test case</h3>
  <dl>
    <dt>Open this test page with Internet Explorer 9 or 11</dt>
    <dd class="ok">2 tables are shown</dd>
    <dt>Type some text in filter input in table 1</dt>
    <dd class="ok">Rows are filtered and the caret is positioned at the end of the text</dd>
    <dt>Type some text in filter input in table 2</dt>
    <dd class="ko">Rows are filtered, but the caret is positioned at the begining of the text 
                  (its position should not be changed)</dd>
    <dt>In table 1, click in the middle of the text and type some new text</dt>
    <dd class="ko">Rows are filtered, but the caret is positioned at the end of the text 
                  (its position should not be changed)</dd>
  </dl>

<hr />

<h2>Table 1 (OK)</h2>
<table id="table-ok">
    <thead>
    <tr>
        <th data-field="name" data-filter-control="input">Name</th>
        <th data-field="forks_count">Forks</th>
    </tr>
    </thead>
</table>

<hr />

<h2>Table 2 (KO)</h2>
<table id="table-ko">
    <thead>
    <tr>
        <th data-field="name" data-filter-control="input">Name</th>
        <th data-field="forks_count">Forks</th>
    </tr>
    </thead>
</table>
```

```javascript
var data = [
    { "name": "alaborum qui modi", "forks_count": "766", },
    { "name": "explicabo et id", "forks_count": "7908", },
    { "name": "libero qui ipsum", "forks_count": "180", },
    { "name": "numquam repellendus voluptatem", "forks_count": "1533", },
    { "name": "eos et tempore", "forks_count": "470", },
    { "name": "distinctio nesciunt amet", "forks_count": "240", },
    { "name": "aut libero saepe", "forks_count": "0", },
    { "name": "quos repudiandae fuga", "forks_count": "861", },
    { "name": "expedita deserunt quibusdam", "forks_count": "671", },
    { "name": "expedita aliquid non", "forks_count": "97145", },
];

$(function () {
    $('#table-ok').bootstrapTable({
        data: data,
        filterControl: true,
    });
    $('#table-ko').bootstrapTable({
        data: data,
        filterControl: true,
        height: 300,
    });
});
```

```css
/* very basic styling for instructions only */
dl { font-size: 14px; margin-top:5px; }
dt:before { content: '* '; }
dd { padding-left: 15px; }
.ok { color: green; }
.ko { color: red; }
pre { display: inline-block; }
```

### LINKED ISSUE #2082: Input text going back to the beggining of filter-control
Hello,

When using bootstrap-table with the filter-control plugin, and you write something on the column filter input text, after the filter is applied the cursor goes back to the beggining of the input element.

This issue happens when using Internet Explorer 11 11.0.9600.17801 (does not happen in chrome).

### LINKED ISSUE #2085: Duplicate data field issue on show hide columns and table rendering
I've one table which is using AJAX data to load and working fine, but i need to use one field multiple times in table to make links, with check-box and more to associate the IDs.
```
 <tr>
        <th data-field="id" data-checkbox="true"></th>
        <th data-field="id" data-sortable="true">Item ID</th>
        <th data-field="id" data-formatter="linkFormatter" data-sortable="true">Item Link</th>
    </tr>
```
But when i use the same field the heading style don't synced properly and the show/hide columns is also broken.

See the screenshot and this JsFiddle :: http://jsfiddle.net/4r6g4cfu/194/

image

Looking for help

### LINKED ISSUE #2091: data-filter-control="select" with Pagination issue
There seems to be an issue using data-filter-control="select" with pagination.  The select only shows values from the initial page.  Additional values are added after selecting additional pages.

1. Select dropdown only show values on page 1
2. click on page 2
3. Select now shows values from page 1 and page 2
4. click on page 3
5. Select now show values from page 1, 2, and 3

fiddle: https://jsfiddle.net/edsoja/cyuqeghz/1/

I believe the issue may be in bootstrap-table-filter-control.js, line 143
`for (var i = bootstrapTable.pageFrom - 1; i < bootstrapTable.pageTo; i++) {`

### LINKED ISSUE #2098: Some problem when search data include "&"
Hi all,

I have some problem about search function.
I make a example to explain this problem easily: https://jsfiddle.net/RJ_Hsiao/dsn41bt0/
You can see that there are some data include "`&`" character in the "Division" column, like "F&SS", "S&O".
In this case, I cannot get any result when I use "`F&SS`" into search input, I have to use "`F&amp;SS`" to get the result I want.

Is there any way to let me pre-process search text?

### LINKED ISSUE #2099: Extend bootstrap-table-cookie with 2nd script
Hi,
I wrote an extension to filter my data. I want to save the state of my extension with bootstrap-table-cookie.
Is there a possibility to access the get and setCookie functions from another script?

I don't want to edit the bootstrap-table-cookie source. I want to write a 2nd script called bootstrap-table-cookie-ext.js

### PR DESCRIPTION
**Pull request by djhvscf**
Fix #1838 
Fix #2005 
Fix #2071 
Fix #2082 
Fix #2085 
Fix #2091 
Fix #2098 
Fix #2099 

Example for #2005: 
- http://jsfiddle.net/4r6g4cfu/215/

Example for #2071 and #2082 :
- in IE: https://jsfiddle.net/7t5tocn7/3/

Example #2085: 
- http://jsfiddle.net/djhvscf/4r6g4cfu/220/

Example for #2091:
- sidePagination client: https://jsfiddle.net/djhvscf/cyuqeghz/5/
- sidePagination server: http://jsfiddle.net/djhvscf/4r6g4cfu/218/

Example for #2098:
- With escape set to false:
- https://jsfiddle.net/djhvscf/dsn41bt0/3/
- With escape set to true:
- https://jsfiddle.net/djhvscf/dsn41bt0/4/

Example for #2099 
- http://jsfiddle.net/djhvscf/e3nk137y/5415/

### CODE REVIEW

File: src/bootstrap-table.js [hunk 1]
```diff
@@ -697,7 +697,7 @@
             return;
         }
 
-        this.fromHtml = true;
+        this.fromHtml = this.options.sidePagination === 'server' ? false : true;
```
Comment (wenzhixin) [inline src/bootstrap-table.js]: `this.fromHtml = this.options.sidePagination === 'client';`
or `this.fromHtml = this.options.sidePagination !== 'server';`

File: src/bootstrap-table.js [hunk 2]
```diff
@@ -2115,8 +2116,12 @@
                 index = i - 1;
             }
 
-            that.$header_.find(sprintf('th[data-field="%s"]', visibleFields[index]))
-                .find('.fht-cell').width($this.innerWidth());
+            var th = that.$header_.find(sprintf('th[data-field="%s"]', visibleFields[index]));
```
Comment (wenzhixin) [inline src/bootstrap-table.js]: it is better to use `$th` instead.

File: src/bootstrap-table.js (hunk 1 above)
Comment (djhvscf) [inline src/bootstrap-table.js]: Done

File: src/bootstrap-table.js (hunk 2 above)
Comment (djhvscf) [inline src/bootstrap-table.js]: Done

**Comment by wenzhixin** [comment]
Great! Thanks very much :)

**Pull request merged**